### PR TITLE
FEATURE: SD-150 - Add new functions to adapt new APIs for exporting

### PR DESCRIPTION
**Pull request by ptquang86**
PR for [SD-150](https://jira.intgdc.com/browse/SD-150)

- Add new ReportModule and exportResult function

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/116/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/139/ : FAILURE in 45s

File: src/interfaces.ts [hunk 1]
```diff
@@ -316,3 +318,12 @@ export interface ISdkOptions {
 export interface IFeatureFlags {
     [key: string]: number | boolean | string;
 }
+
+export type ExportConfigFormat = 'xlsx' | 'csv' | 'raw';
+
+export interface IExportConfig {
+    title?: string;
+    format?: ExportConfigFormat;
+    mergeHeaders?: false | true;
```
Comment (martinmosko) [inline src/interfaces.ts]: I think this could be `boolean` as a type ;)

File: src/interfaces.ts [hunk 2]
```diff
@@ -316,3 +318,12 @@ export interface ISdkOptions {
...
+    includeFilterContext?: false | true;
```
Comment (martinmosko) [inline src/interfaces.ts]: `includeFilterContext: boolean`

File: src/report.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+     * @param {String} projectId GoodData projectId
+     * @param {String} executionResult report which should be exported
+     * @param {IExportConfig} exportConfig requested export options
+     * @return {IExportResponse} uri
+     */
+    public exportResult(
+        projectId: string,
+        executionResult: string,
+        exportConfig: IExportConfig = {}): Promise<IExportResponse> {
+        const data = {
```
Comment (martinmosko) [inline src/report.ts]: Consider naming as `requestData` or `requestPayload` :)

**Comment by ptquang86** [comment]
Hi @martinmosko, all requested changes have been done. Thank you

File: test/report.test.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                    {
+                        status: 200,
+                        body: { uri: createdReport }
+                    }
+                );
+
+                const executionResult = '/executionResult/1234';
+                const exportConfig = {
+                    title: 'title',
+                    format: 'xlsx' as ExportConfigFormat,
```
Comment (martinmosko) [inline test/report.test.ts]: this `as "Re-typing"` is not necessary :) just you have to specify type of whole object

```
const exportConfig: IExportConfig = {
	title: 'title',
    format: 'xlsx',
	mergeHeaders: false,
    includeFilterContext: true
};
```

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/118/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/141/ : SUCCESS in 3m 22s

**Comment by ptquang86** [comment]
Updated. Thanks @martinmosko,

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/119/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/142/ : SUCCESS in 1m 17s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/124/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/148/ : SUCCESS in 34s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/125/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/149/ : FAILURE in 24s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/126/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/151/ : SUCCESS in 40s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/127/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/152/ : SUCCESS in 1m 26s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/128/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/153/ : SUCCESS in 1m 02s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/129/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/154/ : SUCCESS in 1m 35s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/130/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/155/ : SUCCESS in 2m 29s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/131/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/156/ : SUCCESS in 1m 56s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/134/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/160/ : SUCCESS in 2m 50s

File: src/report.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        };
+        return this.xhr.post(`/gdc/internal/projects/${projectId}/exportResult`, { body: requestPayload })
+            .then((response: ApiResponse) => response.getData())
+            .then((data: IExportResponse) =>
+                handlePolling(this.xhr.get.bind(this.xhr), data.uri, this.isDataExported, {}, {
+                    method: 'HEAD'
+                }));
+    }
+
+    private isDataExported(response: any) {
```
Comment (martinmosko) [inline src/report.ts]: function return type and `any`

File: src/report.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            resultExport: {
+                executionResult,
+                exportConfig
+            }
+        };
+        return this.xhr.post(`/gdc/internal/projects/${projectId}/exportResult`, { body: requestPayload })
+            .then((response: ApiResponse) => response.getData())
+            .then((data: IExportResponse) =>
+                handlePolling(this.xhr.get.bind(this.xhr), data.uri, this.isDataExported, {}, {
+                    method: 'HEAD'
```
Comment (martinmosko) [inline src/report.ts]: handlePolling has only 4 arguments not 5...

File: src/report.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+     * @param {String} executionResult report which should be exported
+     * @param {IExportConfig} exportConfig requested export options
+     * @return {Promise} Resolves if export successfully,
+     *                   Reject if export has error (network error, api error)
+     */
+    public exportResult(
+        projectId: string,
+        executionResult: string,
+        exportConfig: IExportConfig = {}): Promise<IExportResponse | Error> {
+        const requestPayload = {
```
Comment (martinmosko) [inline src/report.ts]: missing interface

File: src/report.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            resultExport: {
+                executionResult,
+                exportConfig
+            }
+        };
+        return this.xhr.post(`/gdc/internal/projects/${projectId}/exportResult`, { body: requestPayload })
+            .then((response: ApiResponse) => response.getData())
+            .then((data: IExportResponse) =>
+                handlePolling(this.xhr.get.bind(this.xhr), data.uri, this.isDataExported, {}, {
```
Comment (martinmosko) [inline src/report.ts]: also `options` has default value `{}`

File: test/report.test.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                    projectUri,
+                    {
+                        status: 200,
+                        body: { uri: createdReport }
+                    }
+                );
+
+                const mockTask = (status: any) => ({ status });
+                const finishedTask = mockTask('OK');
+                const runningTask = mockTask('RUNNING');
```
Comment (martinmosko) [inline test/report.test.ts]: looks like mockTask responses do not match `isDataExported` test in handlePooling

File: test/report.test.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                const createdReport = '/gdc/exporter/result/12345';
+                fetchMock.mock(
+                    projectUri,
+                    {
+                        status: 200,
+                        body: { uri: createdReport }
+                    }
+                );
+
+                const mockTask = (status: any) => ({ status });
```
Comment (martinmosko) [inline test/report.test.ts]: `any`

File: test/report.test.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+                const executionResult = '/executionResult/1234';
+                const exportConfig: IExportConfig = {
+                    title: 'title',
+                    format: 'xlsx',
+                    mergeHeaders: false
+                };
+
+                return mockedReportModule().exportResult(projectId, executionResult, exportConfig)
+                    .then((result: any) => expect(result.uri).toEqual(createdReport));
```
Comment (martinmosko) [inline test/report.test.ts]: `any`

File: test/report.test.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+            it('should reject with 400 when resource fails', () => {
+                fetchMock.mock(
+                    projectUri,
+                    400
+                );
+
+                const executionResult = '/executionResult/1234';
+                return mockedReportModule().exportResult(projectId, executionResult)
+                    .then(null, (error: any) => expect(error).toBeInstanceOf(Error));
```
Comment (martinmosko) [inline test/report.test.ts]: `any`

File: src/util.ts [hunk 13]
```diff
@@ -39,33 +39,37 @@ export interface IPollingOptions {
...
         maxAttempts = 50,
         pollStep = 5000
     } = options;
 
-    return xhrGet(uri)
-        .then((r: any) => r.getData())
+    const isHeadRequest = settings.method === 'HEAD';
+
+    return xhrGet(uri, settings)
+        .then((r: any) => isHeadRequest ? r.getHeaders() : r.getData())
```
Comment (martinmosko) [inline src/util.ts]: what is the purpose of this code ?

File: src/util.ts [hunk 14]
```diff
@@ -39,33 +39,37 @@ export interface IPollingOptions {
...
+ * @param {Object} settings for xhr request (method)
  * @private
  */
 export const handlePolling = (
     xhrGet: any,
     uri: string,
     isPollingDone: Function,
-    options: IPollingOptions = {}
+    options: IPollingOptions = {},
+    settings: any = {}
```
Comment (martinmosko) [inline src/util.ts]: this looks suspicious... what is the purpose of this change?

Comment (martinmosko) [reply src/util.ts]: well then I would say we should change the first argument and not add a settings like this :)

**Comment by ptquang86** [comment]
Thanks @martinmosko

File: src/report.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
+// (C) 2007-2018 GoodData Corporation
+import { XhrModule, ApiResponse } from './xhr';
+import { IExportConfig, IExportResponse } from './interfaces';
+import { handlePolling } from './util';
+
+/**
+ * Functions for working with projects
```
Comment (BugsBunny338) [inline src/report.ts]: Functions for working with reports

File: src/report.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @class project
```
Comment (BugsBunny338) [inline src/report.ts]: report

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/146/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/173/ : SUCCESS in 2m 27s

**Comment by ptquang86** [comment]
Hi @martinmosko, @BugsBunny338, I have updated this PR based on your comments. 
Please review again. 
Thanks

File: test/report.test.ts [hunk 17]
```diff
@@ -28,9 +28,9 @@ describe('report', () => {
                     }
                 );
 
-                const mockTask = (status: any) => ({ status });
-                const finishedTask = mockTask('OK');
-                const runningTask = mockTask('RUNNING');
+                const mockTask = (status: number) => ({ status, uri: createdReport });
+                const finishedTask = mockTask(200);
```
Comment (martinmosko) [inline test/report.test.ts]: also include please a test-case with pooling and status-code 400 :)

File: src/report.ts [hunk 18]
```diff
@@ -0,0 +1,59 @@
+// (C) 2007-2018 GoodData Corporation
+import { XhrModule, ApiResponse } from './xhr';
+import { IExportConfig, IExportPromise, IExportResponse } from './interfaces';
+import { handlePolling, IPollingOptions } from './util';
+
+interface IResultExport {
```
Comment (martinmosko) [inline src/report.ts]: I would move the interfaces to interfaces file next to others as `IExportConfig` :) which is already there... what do you think?

Comment (ptquang86) [reply src/report.ts]: IResultExport and IExportResultPayload are used internally in report.ts.
If I move them to interfaces file, they will be exported outside too

// gooddata-browser.ts
export * from './interfaces';

In my opinion, I prefer to leave them there

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by ptquang86** [comment]
Hi @martinmosko, please review again. Thank you

File: src/report.ts (hunk 18 above)
Comment (martinmosko) [reply src/report.ts]: ok :)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-opensource-repository-check/158/ : SUCCESS in 13s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/188/ : SUCCESS in 55s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-merge-compliance-check/50/ : SUCCESS in 25s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/201/ : SUCCESS in 1m 14s
